apiserver/facades/controller/caasapplicationprovisioner/provisioner.go

```diff
@@ -120,14 +126,23 @@ func NewStateCAASApplicationProvisionerAPI(ctx facade.Context) (*APIGroup, error
 		PasswordChanger:          common.NewPasswordChanger(st, common.AuthFuncForTagKind(names.ApplicationTagKind)),
 		LifeGetter:               common.NewLifeGetter(st, common.AuthFuncForTagKind(names.ApplicationTagKind)),
 		AgentEntityWatcher:       common.NewAgentEntityWatcher(st, resources, common.AuthFuncForTagKind(names.ApplicationTagKind)),
-		CharmsAPI:                commonCharmsAPI,
-		ApplicationWatcherFacade: common.NewApplicationWatcherFacadeFromState(st, resources, common.ApplicationFilterCAASSidecar),
+		charmInfoAPI:             commonCharmsAPI,
+		appCharmInfoAPI:          appCharmInfoAPI,
+		ApplicationWatcherFacade: common.NewApplicationWatcherFacadeFromState(st, resources, common.ApplicationFilterNone),
 		API:                      api,
 	}
 
 	return apiGroup, nil
 }
 
+func (a *APIGroup) CharmInfo(args params.CharmURL) (params.Charm, error) {
```

Comment:
It is composition, right? (Go doesn't have inheritance.) I guess you mean an [embedded field](https://golang.org/ref/spec#Struct_types). I actually had these as embedded fields originally, but Joe [commented above](https://github.com/juju/juju/pull/13090#discussion_r665300199) that we've moved away from this as "it makes versioning a nightmare" -- I guess mainly because you're not pulling in the methods explicitly, so if someone adds a method to the common type, the type embedding it magically has another method, so you'd have to run around check all of the uses of the common type. So I switched to non-embedded named fields for the stuff I'm adding, and wrapped the methods explicitly.